### PR TITLE
Update from development of the CLI for the PYRAMID project at Newcastle University

### PR DESCRIPTION
**Pull request by Robin-Wardle**
Main changes:

- Changed the authentication process to use keycloak
- Fixed the endpoints for adding, deleting and updating a Model
- Added a Workflow object for adding, deleting and updating a Workflow
- Modified some of the documentation to more accurately reflect installation process on Linux vs Windows
- Included some example outputs for Models and Workflows, for reference
- Many comments and docstrings updated

The structure of the Model and Workflow objects have changed because of the way in which Metadata is handled by the DAFNI API. There is no longer a /metadata/ component to API calls and metadata is instead included directly in the JSON object returned from the server.

NOT implemented
- full test suite for the new / changed functions
- not corrected the dataset interface endpoints

There is still a lot of WIP code (commented out) for eventual review / development / deletion

### CODE REVIEW

**Review by sarahbyrnie — APPROVED** [review]

File: dafni_cli/workflow/version_history.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        ):
+            latest_version.get_details_from_id(jwt_string, latest_version.version_id)
+
+        self.dictionary = latest_version.dictionary["version_history"]
+        self.history = [latest_version]
+
+        if len(self.dictionary) > 1:
+            for version_dict in self.dictionary[1:]:
+                version = Workflow()
+                version.get_details_from_id(jwt_string, version_dict["id"])
```
[the remark this answers is not in the document]
Comment (Robin-Wardle) [reply dafni_cli/workflow/version_history.py]: I fixed this in the repo before I saw this correction.

File: dafni_cli/workflow/workflow.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+from dafni_cli.utils import (
+    prose_print,
+    print_json
+)
+from dafni_cli.auth import Auth
+
+
+class Workflow:
+    """
+    Contains information about a DAFNI workflow DAFNI.
```
Comment (roseDickinson) [inline dafni_cli/workflow/workflow.py:21]:
```suggestion
    Contains information about a DAFNI workflow.
```

File: dafni_cli/workflow/workflow.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+        get_metadata(jwt (str)): After details have been obtained, populate metadata attributes.
+        filter_by_date(key (str), date (str)): calculates whether the workflow was created/published before a date.
+        output_details(): Prints key information of workflow to console.
+        output_metadata(): Prints key information of workflow metadata to console.
+
+    Attributes:
+        api_version: The version of the API that created the data
+        auth: credentials used to retrieve the data
+        creation_date: Date and time the workflow was created
+        description: More detailed information on the model
```
Comment (roseDickinson) [inline dafni_cli/workflow/workflow.py:36]:
```suggestion
        description: More detailed information on the workflow
```

File: dafni_cli/workflow/workflow.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+        display_name: Name of the workflow used in the web app
+        id: ID used to identify the specific workflow version
+        instances: A list of the workflow version run instances
+        kind: The type of asset (will be 'W', workflow)
+        metadata: a sub-group containing the workflow metadata:
+            description: More detailed information on the model
```
Comment (roseDickinson) [inline dafni_cli/workflow/workflow.py:42]:
```suggestion
            description: More detailed information on the workflow
```

File: dafni_cli/workflow/workflow.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+            display_name: Name of the workflow used in the web app
+            name: DAFNI workflow name
+            publisher: Organisation publishing the model
```
Comment (roseDickinson) [inline dafni_cli/workflow/workflow.py:45]:
```suggestion
            publisher: Organisation publishing the workflow
```

File: dafni_cli/workflow/workflow.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+            summary: One-line summary of what the model does
```
Comment (roseDickinson) [inline dafni_cli/workflow/workflow.py:46]:
```suggestion
            summary: One-line summary of what the workflow does
```

File: dafni_cli/workflow/workflow.py [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+            source_code: Generally, a link (e.g. Github) link to the project source
```
Comment (roseDickinson) [inline dafni_cli/workflow/workflow.py:47]:
```suggestion
```
source code doesn't exist for workflows

File: dafni_cli/workflow/workflow.py [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+        publication_date: Date and time the model was published
+        publisher: Organisation publishing the model
+        spec: JSON description of the workflow steps
+        summary: One-line summary of what the model does
+        version_history: A list of all versions of the workflow
+        version_message: Message attached when the model was updated to this model version
+        version_tags: Any tags created by the publisher for this version
+
+        dictionary: Full description of the model as retrieved from the API
+        metadata_obj: ModelMetadata object containing metadata for the model
```
Comment (roseDickinson) [inline dafni_cli/workflow/workflow.py:62]:
```suggestion
        publication_date: Date and time the workflow was published
        publisher: Organisation publishing the workflow
        spec: JSON description of the workflow steps
        summary: One-line summary of what the workflow does
        version_history: A list of all versions of the workflow
        version_message: Message attached when the workflow was updated to this workflow version
        version_tags: Any tags created by the publisher for this version

        dictionary: Full description of the workflow as retrieved from the API
        metadata_obj: WorkflowMetadata object containing metadata for the workflow
```

File: dafni_cli/workflow/workflow.py [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+
+
+    def filter_by_date(self, key: str, date: str) -> bool:
+        """Filters workflows based on the date given as an option.
+        Args:
+            key (str): Key for WORKFLOW_DICT in which date is contained
+            date (str): Date for which workflows are to be filtered on: format DD/MM/YYYY
+
+        Returns:
+            bool: Whether to display the model based on the filter
```
Comment (roseDickinson) [inline dafni_cli/workflow/workflow.py:178]:
```suggestion
            bool: Whether to display the workflow based on the filter
```

File: dafni_cli/workflow/workflow.py [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+        return missing_workflow_attributes
+
+
+    def get_attributes_from_id(self, jwt_string: str, version_id_string: str):
+        """
+        Retrieve workflow attributes from the DAFNI API using the /workflows/<version-id> endpoint.
+ 
+        Args:
+            jwt_string (str): JWT for login purposes
+            version_id_string (str): Version ID of the model
```
Comment (roseDickinson) [inline dafni_cli/workflow/workflow.py:151]:
```suggestion
            version_id_string (str): Version ID of the workflow
```

**Review by roseDickinson — APPROVED** [review]

**Pull request merged**